### PR TITLE
Create a class representing a map from ModuleCode to RoleType

### LINKED ISSUE #72: Create a class representing a map from `ModuleCode` to `RoleType`
This class will contain a `HashMap<ModuleCode, RoleType>`. It represents a map from `ModuleCode` to `RoleType` for a given Person.

This issue is just about writing code for the class.

### PR DESCRIPTION
**Pull request by MAOXIONGKAI**
Closes #72 

### APIs implemented in Role.java class are as such:
1. Constructors(overloaded to suit different implementation requirements, both keeps the class immutable)
    1.1 First one takes in two arrays(Module codes and role types respectively), validates and then creates a HashMap<ModuleCode, RoleType> for the Role object, this is helpful when creating new Role for a new Person based on the given add command.
    1.2 Second one takes in a  HashMap<ModuleCode, RoleType> directly, this is helpful when updating/editing a Person's particulars.
2. areValidRoleTypes(String[] roleTypes): check if a given role type String array contains all valid role keywords, used by constructor 1.1
3. roleTypeStringToEnum(String keyword): converts a valid role type String keyword to its corresponding RoleType Enum value
4. getRoleDescription(RoleType roleType): provides an intuitive String description of the roles taken by the Person that are of the given role type, this is used by toString() method
5. toString(), equals(), hashCode() - for testing purpose

### Regarding the test cases
1. The test cases are designed to cover all the above methods with common inputs and common edge cases, feel free to suggest any improvements if needed, thanks
2. Take note that in order for the test cases in this PR to work properly, PR #91 has to be merged first because some of the test on toString() and getRoleDescription() are based on these changes.

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Role.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    /**
+     * Default constructor for a {@code Role}.
+     * This is for creating roles for a Person based on the module code and role types
+     * extracted from an add command.
+     *
+     * @param moduleCodes Array of modules codes
+     * @param roleTypes Array of role type String
+     */
+    public Role(ModuleCode[] moduleCodes, String[] roleTypes) {
+        requireNonNull(moduleCodes);
```
Comment (weiseng18) [inline src/main/java/seedu/address/model/person/Role.java:31]: hmm do you want a constructor of `Role(ModuleCode[] moduleCodes, RoleType[] roleTypes)` instead? assume that this class will receive the `RoleType[]`, then the conversion of user input of type `String` to the `RoleType` will be done in the parser

File: src/main/java/seedu/address/model/person/Role.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        }
+    }
+
+    /**
+     * Get the String representation of a Person's Role for a specific
+     * role type.
+     *
+     * @param roleType Type of role to search and display.
+     * @return String representation of the roles with the specified type.
+     */
```
Comment (weiseng18) [inline src/main/java/seedu/address/model/person/Role.java:102]: nice idea!

File: src/main/java/seedu/address/model/person/Role.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    public String getRoleDescription(RoleType roleType) {
+        requireNonNull(roleType);
+        String prefix = roleType.toString() + " of: ";
+        StringBuilder description = new StringBuilder(prefix);
+        this.roles.entrySet().stream()
+                .filter(role -> role.getValue().equals(roleType))
+                .map(Map.Entry::getKey)
```
Comment (weiseng18) [inline src/main/java/seedu/address/model/person/Role.java:109]: do you want to extract this part out to a `private` function in this class? i think it might improve readability

then you can use `description.append(String.join(", ", list))` to avoid dealing with the trailing comma

**Review by weiseng18 — COMMENTED** [review]
great job with adding nontrivial new code for our MVP!

i think for the constructor, since we have a parser, it might be ok to assume that this class should receive a `RoleType[]` array, just like how in the other constructor, it receives a `HashMap` that has `RoleType[]`

the rest of the comments are just an opinion, up to you if you want to make those changes

File: src/main/java/seedu/address/model/person/Role.java (hunk 2 above)
Comment (MAOXIONGKAI) [reply src/main/java/seedu/address/model/person/Role.java]: Thank you! I have included an example in the Javadoc for easier understanding!

File: src/main/java/seedu/address/model/person/Role.java (hunk 3 above)
Comment (MAOXIONGKAI) [reply src/main/java/seedu/address/model/person/Role.java]: Thanks for the suggestion! I have extracted out this part of the logic and create two new methods `getFilteredModuleCodes(RoleType)` and `getModuleCodeString(Stream<ModuleCode>)` to improve the overall readability of the code and provide more useful APIs for future development 🎉

File: src/main/java/seedu/address/model/person/Role.java (hunk 1 above)
Comment (MAOXIONGKAI) [reply src/main/java/seedu/address/model/person/Role.java:32]: Thanks for the suggestion! I realized that all other classes that represents attribute of a Person has validation steps at the parsing command, object creation and loading JSON data stage. For the Role class, since the data passed in during parsing phase and the actual data stored are of different types, I think if we are to implement validations are all three stages we need multiple validation methods to handle different types of data. My thoughts are like these:
1. For parsing stage, the parser just checks if the module code and the roles keywords are valid inputs and then convert them to respective ModuleCode[] and RoleType[] array and pass these two array to the constrcutor
2. For constructor itself, we can validate the inputs by checking if there's any null object in the given array
3. For `JsonAdaptedRole`(should be implemented for save/load feature), we also check if there's any null objects and then convert them to their respective Model type.
Do you think this approach is sufficient to make sure that the program won't break due to any invalid input or messing up of the storage? Thank you for your time

Comment (weiseng18) [reply src/main/java/seedu/address/model/person/Role.java:32]: yep I agree with your thoughts! so just to paraphrase, there should be a

- `String -> RoleType` convert + validity check
- `RoleType` non-null check
- `ModuleCode` non-null check

1. Parsing stage uses all 3 (similar to `Tag`)
2. Constructor uses non-null checkers only because parsing stage will return `RoleType[]`, so it's not possible to check the `String` anymore. I'm not sure why the provided `Tag` class runs its validity checker both at the parser and the constructor
3. `JsonAdaptedRole` should use all 3, since it is basically doing parsing and constructing. However, it seems like `JsonAdaptedTag` does not check for `null` at all, which is a bug in AB3

File: src/main/java/seedu/address/model/person/ModuleRoleMap.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    public ModuleRoleMap(ModuleCode[] moduleCodes, RoleType[] roleTypes) {
+        requireNonNull(moduleCodes);
+        requireNonNull(roleTypes);
+
+        if (moduleCodes.length != roleTypes.length) {
+            throw new IllegalArgumentException(
+                    "Module code array's length is not consistent with that of role type array.");
+        }
+
+        checkArgument(isValidModuleCodeArray(moduleCodes), ModuleCode.MESSAGE_CONSTRAINTS);
```
Comment (huangjiaxi1111) [inline src/main/java/seedu/address/model/person/ModuleRoleMap.java:49]: Is it possible to write another method to check for null object in an array(to avoid duplicate codes)?
Also, the MESSAGE_CONSTRAINTS I think should be changed to some other error message like "Null object not allowed" since this essentially checks for that. (though Im not sure under what circumstances will there be null object)

File: src/main/java/seedu/address/model/person/ModuleRoleMap.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    /**
+     * Returns true if the given role type String array contains all
+     * valid role type keywords.
+     *
+     * @param roleTypes Array of role type keywords
+     * @return whether the array contains all valid keywords
+     */
+    public static boolean areValidRoleTypeKeywords(String[] roleTypes) {
```
Comment (huangjiaxi1111) [inline src/main/java/seedu/address/model/person/ModuleRoleMap.java:77]: I think it makes more sense to move this to RoleType class.
Same for roleTypeStringToEnum

File: src/main/java/seedu/address/model/person/ModuleRoleMap.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+     *
+     * @param moduleCodeStream filtered Stream of module codes.
+     * @return comma separated String representation of a list of ModuleCode
+     */
+    public static String getModuleCodeString(Stream<ModuleCode> moduleCodeStream) {
+        requireNonNull(moduleCodeStream);
+        String[] moduleCodeStringArray = moduleCodeStream
+                .map(ModuleCode::toString)
+                .toArray(String[]::new);
+        return String.join(", ", moduleCodeStringArray);
```
Comment (huangjiaxi1111) [inline src/main/java/seedu/address/model/person/ModuleRoleMap.java:168]: Can be simplified as such:
```
return moduleCodeStream
            .map(ModuleCode::toString)
            .collect(Collectors.joining(", "));
```

File: src/main/java/seedu/address/model/person/ModuleRoleMap.java (hunk 4 above)
Comment (MAOXIONGKAI) [reply src/main/java/seedu/address/model/person/ModuleRoleMap.java:49]: Hi, I agree with the suggestion on avoiding code duplication, but regarding `MESSAGE_CONSTRAINTS` I think they are actually displayed to the users, and telling them arrays contains null objects are not very useful and instead we should tell them how to input data properly to avoid null objects in the generated arrays. Would you think it would be better to leave the MESSAGE_CONSTRAINTS as it is?

File: src/main/java/seedu/address/model/person/ModuleRoleMap.java (hunk 6 above)
Comment (MAOXIONGKAI) [reply src/main/java/seedu/address/model/person/ModuleRoleMap.java:168]: Thanks for the suggestion! Will implement the change! Thank you!

**Comment by MAOXIONGKAI** [comment]
Implemented new changes according to @huangjiaxi1111 's suggestions. These changes are finished with the help of @lithops-zty.
The changes are as such:
1. Remove `isValidModuleCodeArray` and `isValidRoleTypeArray` and replace them with `requireAllNonNull` methods provided by `CollectionUtil` class. However, since not all overloaded version of `requireAllNonNull` matches our need(since for one of the constructors we need check on a HashMap instead of array or ArrayList), we have implemented an overloaded  version of `requireAllNonNull` that accepts in a HashMap, this may be useful for some of you guys in the future.
2. Remove `convertStringtoEnum` method from `ModuleRoleMap` class. However, instead of moving this to `RoleType` class as suggested by Jiaxi, me and Tianyi think this functionality actually belongs to `ParserUtil` class(Since it accepts user input and convert it to some other data), therefore we have implemented a similar method called `parseRoleType` in `ParserUtil` class.
3. Remove `areValidRoleTypeKeywords` method from `ModuleRoleMap` class. However, since the new `parseRoleType` method that we have implemented in `ParserUtil` class takes in a single keyword instead of an array, we decide to change it to `isValidRoleTypeKeyword` instead, because it is better to check the keywords one by one and alert the user about the wrong input immediately instead of checking the whole array and then tell user that the whole array is not accepted because some of the elements are invalid.

File: src/main/java/seedu/address/model/person/ModuleRoleMap.java (hunk 4 above)
Comment (huangjiaxi1111) [reply src/main/java/seedu/address/model/person/ModuleRoleMap.java]: Make sense! Since add module not implemented yet, we are not sure what kind of input can cause this to happen

**Comment by lithops-zty** [comment]
Extremely hard work to get the overloaded check methods work, kudos to @MAOXIONGKAI!👍

**Review by huangjiaxi1111 — APPROVED** [review]
LGTM ! 👍
